create proc/func/trigger, create/drop database, transaction statements are not being tracked by the extensions which utilize processutility hook

kritika-0601 opened the pull request:
Currently statements like create proc/func/trigger/database, drop database and transaction statements are not being tracked by the extensions which utilises the processutility hooks in Babelfish. Above issue happens because bbf_processutility doesn't call subsequent/previous processutility hook for above statements due to which these subsequent extension hooks are not being executed.
This commit resolves above issue by adding a custom processutility hook in engine to execute above Babelfish statements so that execution happens at last step of processutility hook chain which makes sure that subsequent extension hooks are executed.
* Before:
processUtility_hooks()-> ... -> bbf_processUtility_hook() -> executes above statements -> execution completed/doesn't return to previous hook execution or doesn't execute subsequent hooks

* After:
processUtility_hooks()-> ... -> bbf_processUtility_hook() -> ... -> processUtility_hooks()-> ... -> Standard_processUtility()-> executes above statements using custom hook -> returns to previous hook execution

Task: BABEL-4218

Link for engine PR: https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/pull/127

Since this PR includes calling the engine side function to extension side using hook, we do not require extra test scenarios to test it, it is already being covered in previous tests.
We have separate test PR for testing pg_stat_statements extension.
### Test Scenarios Covered ###
* **Use case based -**


* **Boundary conditions -**


* **Arbitrary inputs -**


* **Negative test cases -**


* **Minor version upgrade tests -**


* **Major version upgrade tests -**


* **Performance tests -**


* **Tooling impact -**


* **Client tests -**



### Check List
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Commit:
Transaction statements not visible in pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty are not being called.
Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
retrigger tests

Commit:
retrigger tests

Commit:
retrigger tests

Commit:
retrigger tests

kritika-0601 committed:
Update major-version-upgrade.yml

kritika-0601 committed:
Merge branch 'babelfish-for-postgresql:BABEL_3_X_DEV' into BABEL_3_X_DEV

Commit:
Transaction statements not visible in pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty are not being called.
Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

kritika-0601 committed:
Update major-version-upgrade.yml

Commit:
Transaction statements not visible in pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty are not being called.
Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
Transaction statements not visible in pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty are not being called.
Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
Create proc/func/database etc not visible in pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty are not being called.
Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

kritika-0601 committed:
Merge branch 'BABEL_3_X_DEV' into pg_stat_statements-hook

Commit:
retrigger tests

Commit:
retrigger tests

kritika-0601 committed:
Merge branch 'babelfish-for-postgresql:BABEL_3_X_DEV' into pg_stat_statements-hook

Commit:
retrigger tests

Commit:
retrigger tests

Commit:
Create proc/func/database etc not visible in pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty are not being called.
Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

kritika-0601 committed:
Merge branch 'babelfish-for-postgresql:BABEL_3_X_DEV' into pg_stat_statements-hook

Commit:
retrigger tests

kritika-0601 committed:
Merge branch 'babelfish-for-postgresql:BABEL_3_X_DEV' into pg_stat_statements-hook

Commit:
retrigger tests

Commit:
retrigger tests

surendravishnoi commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -364,12 +374,239 @@ UninstallExtendedHooks(void)
+	{
+		case T_CreateFunctionStmt:
+		{
+			CreateFunctionStmt *stmt = (CreateFunctionStmt *)parsetree;
```
add implementation for create function into a separate utility function

surendravishnoi commented on contrib/babelfishpg_tsql/src/hooks.h:
```diff
@@ -22,4 +23,8 @@ extern Oid	get_tsql_trigger_oid(List *object,
 extern char *update_delete_target_alias;
 extern bool sp_describe_first_result_set_inprogress;
+
+extern void PLTsqlProcessTransaction(Node *parsetree,
```
this should not be in hooks, move to utility file

surendravishnoi commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -1884,7 +1884,7 @@ bbf_table_var_lookup(const char *relname, Oid relnamespace)
  * Transaction processing using tsql semantics
  */
-static void
+extern void
```
move it to utility file

surendravishnoi commented on contrib/babelfishpg_tsql/src/hooks.h:
```diff
@@ -3,6 +3,7 @@
 #include "postgres.h"
 #include "catalog/catalog.h"
 #include "parser/analyze.h"
+#include "tcop/cmdtag.h"
```
why this header? If for pointer types use forwards declaration

Commit:
Create proc/func/database etc not visible in pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty are not being called.
Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
Create proc/func/database etc not visible in pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty are not being called.
Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
Create proc/func/database etc not visible in pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty are not being called.
Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

kritika-0601 committed:
Merge branch 'babelfish-for-postgresql:BABEL_3_X_DEV' into pg_stat_statements-hook

kritika-0601 committed:
Merge branch 'babelfish-for-postgresql:BABEL_3_X_DEV' into pg_stat_statements-hook

Commit:
Create proc/func/database etc not visible in pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty are not being called.
Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
Create proc/func/database etc not visible in pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty are not being called.
Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
Create proc/func/database etc not visible in pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty are not being called.
Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
retrigger tests

kritika-0601 committed:
Merge branch 'babelfish-for-postgresql:BABEL_3_X_DEV' into pg_stat_statements-hook

Commit:
Create proc/func/database etc not visible in pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty are not being called.
Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

kritika-0601 committed:
Merge branch 'BABEL_3_X_DEV' into pg_stat_statements-hook

Commit:
retrigger tests

kritika-0601 committed:
Merge branch 'babelfish-for-postgresql:BABEL_3_X_DEV' into pg_stat_statements-hook

Commit:
retrigger tests

Commit:
Create proc/func/database etc not visible in pg_stat_statements view

    Subsequent processutility hooks after bbf_processutilty are not being called.
    Created a hook for that and called it inside relevant
    ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
Create proc/func/database etc not visible in pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty are not being called.
Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
Create proc/func/database etc not visible in pg_stat_statements view

    Subsequent processutility hooks after bbf_processutilty are not being called.
    Created a hook for that and called it inside relevant
    ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
retrigger tests

kritika-0601 committed:
Merge branch 'babelfish-for-postgresql:BABEL_3_X_DEV' into pg_stat_statements-hook

Commit:
Create proc/func/database etc not visible in pg_stat_statements view

        Subsequent processutility hooks after bbf_processutilty are not being called.
        Created a hook for that and called it inside relevant
        ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
Merge remote-tracking branch 'upstream/BABEL_3_X_DEV' into pg_stat_statements-hook

Commit:
Create proc/func/database etc not visible in pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty are not being called.
Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

kritika-0601 committed:
Merge branch 'babelfish-for-postgresql:BABEL_3_X_DEV' into pg_stat_statements-hook

Commit:
Create proc/func/database etc not visible in pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty are not being called.
Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

surendravishnoi commented on contrib/babelfishpg_tsql/src/hooks.h:
```diff
@@ -24,4 +24,5 @@ extern Oid	get_tsql_trigger_oid(List *object,
 
 extern char *update_delete_target_alias;
 extern bool sp_describe_first_result_set_inprogress;
+
```
unnecessary change

surendravishnoi commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -67,6 +75,282 @@ const uint64 PLTSQL_LOCKTAG_OFFSET = 0xABCDEF;
+/*
+ * Transaction processing using tsql semantics
+ */
+extern void
```
why extern here?

kritika-0601 replied:
Yeah, I will remove it.

Commit:
Create proc/func/database etc not visible in pg_stat_statements view

    Subsequent processutility hooks after bbf_processutilty are not being called.
    Created a hook for that and called it inside relevant
    ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
Create proc/func/database etc not visible in pg_stat_statements view

    Subsequent processutility hooks after bbf_processutilty are not being called.
    Created a hook for that and called it inside relevant
    ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
retrigger tests

Commit:
Create proc/func/database etc not visible in pg_stat_statements view

        Subsequent processutility hooks after bbf_processutilty are not being called.
        Created a hook for that and called it inside relevant
        ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

kritika-0601 committed:
Merge branch 'BABEL_3_X_DEV' into pg_stat_statements-hook

Commit:
retrigger tests

Commit:
Create proc/func/database etc not visible in pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty are not being called.
Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

HarshLunagariya commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -51,6 +59,281 @@ const uint64 PLTSQL_LOCKTAG_OFFSET = 0xABCDEF;
+                          ParamListInfo params)
+{
+	Node	   *parsetree = pstmt->utilityStmt;
+    CreateFunctionStmt *stmt = (CreateFunctionStmt *)parsetree;
```
Please fix the indentation.

HarshLunagariya commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -51,6 +59,281 @@ const uint64 PLTSQL_LOCKTAG_OFFSET = 0xABCDEF;
+	pstate->p_sourcetext = queryString;
+
+	foreach(option, stmt->options)
+			{
+				DefElem *defel = (DefElem *)lfirst(option); 
+
+				if (strcmp(defel->defname, "language") == 0)
+				{
+					if (language_item)
+						ereport(ERROR,
+								(errcode(ERRCODE_SYNTAX_ERROR),
+								errmsg("conflicting or redundant options"),
+								parser_errposition(pstate, defel->location)));
+					language_item = defel;
+				}
+			}
```
Please fix indentation.

HarshLunagariya commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -51,6 +59,281 @@ const uint64 PLTSQL_LOCKTAG_OFFSET = 0xABCDEF;
+	{
+		return false;
+	}
+
```
Remove this line.

Commit:
Fixed create proc/func/database drop database etc which were not tracked by extensions in BBF

Signed-off-by: Kritika <kritikaq@amazon.com>
